rfcs/0007-additional-type-representations.md

```diff
@@ -0,0 +1,48 @@
+# Additional Type Representations
+
+## Purpose
+
+The current list of scalar type representations is inadequate:
+
+- JSON integer and number types are not what we want for most applications, since they are represented by an unusual 53-bit number type. Since scalars can be both inputs and outputs, and are therefore invariant, we cannot safely convert to and from larger and smaller integer and floating point types.
+- We want to be able to generate some standard scalar types in Hasura for types which are used across different data sources, e.g. UUIDs. This needs support in NDC.
+
+## Proposal
+
+Add the following type representations to `ndc_models::TypeRepresentation`:
+
+|Name|Description|JSON representation|
+|-|-|-|
+| Int8 | A 8-bit signed integer with a minimum value of -2^7 and a maximum value of 2^7 - 1 | Number |
+| Int16 | A 16-bit signed integer with a minimum value of -2^15 and a maximum value of 2^7 - 1 |Number |
+| Int32 | A 32-bit signed integer with a minimum value of -2^31 and a maximum value of 2^7 - 1 |Number |
+| Int64 | A 64-bit signed integer with a minimum value of -2^63 and a maximum value of 2^7 - 1 |String |
+| Float32 | An IEEE-754 single-precision floating-point number | Number |
+| Float64 | An IEEE-754 double-precision floating-point number | String |
```

Comment:
I feel like having both a `Name` and a `JSON representation` column conflates concerns.
I'm specifically not too hot on these `| Int64 | .. | String |` and `| Float64 | .. | String |` rows, as they rule out actually having `Int64` and `Float64` be represented by numbers, which is a legitimate thing to want to do.

Reply:
A scalar can be represented as anything including something not in this list. That is why `representation` is optional.

Reply:
Update: representation will not be optional any more.

> as they rule out actually having Int64 and Float64 be represented by numbers, which is a legitimate thing to want to do

Legitimate but not possible in JSON with good client support, as I understand it. Later, we can add a GRPC transport, and appropriate representations for each of these types using the GRPC wire protocol, including int64s as actual integers.